### DISCUSSION TITLE
AOC 2022 Day Four

### DISCUSSION THREAD

**Discussion by doleraj** [Virtual-Coffee/virtualcoffee.io#759] [Advent of Code 2022]
Discussion of solutions for [Advent of Code Day 4](https://adventofcode.com/2022/day/4) here

---
VC Advent of code leaderboard
Go to [leaderboard/private](https://adventofcode.com/2022/leaderboard/private) and use join code 363496-e6f42283

**Comment by doleraj** [comment]
I have done ... a _lot_ of geospatial work in my day. I think I need at least both hands if not two hands and a foot to count the number of times I've had to rewrite range inclusion/intersection logic.

So I did this functional-style, but slow, because I have a lot of other crap to do today! 
https://github.com/doleraj/adventofcode/blob/main/aoc2022/src/day04/index.ts

**Reply by tkshill** [reply]
This is awesome! Did you notice any major difference performance-wise?

**Reply by doleraj** [reply]
Faster than yesterday's by quite a bit, but that's probably because number comparisons are a lot faster than string lookups.

**Comment by clandau** [comment]
JS: https://github.com/clandau/advent-of-code/blob/main/2022/day-4/index.js

**Reply by bogdaaamn** [reply]
```js
const pair = pairs.split(',');
const pair1 = pair[0].split('-').map(v => parseInt(v));
const pair2 = pair[1].split('-').map(v => parseInt(v));
```

I spent an EMBARRASSING amount of time debugging just because I was using an array of strings instead of integers 😮‍💨 Although on the test example the string worked fine 😂

**Reply by Cerchie** [reply]
^^^^ same here @BogDAAAMN ! I used different methods but they were treating the types differently.

**Comment by danieltott** [comment]
Here's my ts solution!

https://github.com/danieltott/aoc2022/blob/main/src/day04/index.ts

I also did a pull request this time to see the steps and stuff. I wanted a way to save the state of the part one solution, but it turns out I didn't really need to change anything this time for part 2. Still I think it's kinda fun (: 
https://github.com/danieltott/aoc2022/pull/1

**Reply by clandau** [reply]
ooo i like this:
`    const [aStart, aStop, bStart, bStop] = row.split(/[,-]/).map(Number);`
I forget you can do that with arrays too. 
much cleaner

**Reply by tkshill** [reply]
Agreed with @clandau, `const [aStart, aStop, bStart, bStop] = row.split(/[,-]/).map(Number)` is such a powerful one-liner

**Reply by bogdaaamn** [reply]
WOW this is some cool magic!! I went the classic route

```js
  // Split line into two the two ranges
  const ranges = line.split(",");

  const firstElf = ranges[0].split("-").map((n) => parseInt(n));
  const secondElf = ranges[1].split("-").map((n) => parseInt(n));
```

**Reply by bogdaaamn** [reply]
Is `.map(Number)` a typescript thing or can I use it in js as well? It look much better than `.map((n) => parseInt(n));`

**Reply by tkshill** [reply]
@BogDAAAMN its js. There's also Boolean. They're explicit casts into the data type. As a general rule, if it's Doing a thing, it's js

**Reply by bogdaaamn** [reply]
hmmm so it's sort of a constructor for the data type? because I saw uppercase and I thought that's not very javascript of you 😂

**Reply by clandau** [reply]
totally missed the map to `Number`, had to look that one up, it's a cool trick if you know they're going to be ints: ([source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)):
you also don't have to explicitly pass the function 🤔 
```
// Same as above, but using the concise arrow function syntax
["1", "2", "3"].map((str) => parseInt(str)); // [1, 2, 3]

// A simpler way to achieve the above, while avoiding the "gotcha":
["1", "2", "3"].map(Number); // [1, 2, 3]

// But unlike parseInt(), Number() will also return a float or (resolved) exponential notation:
["1.1", "2.2e2", "3e300"].map(Number); // [1.1, 220, 3e+300]
```

**Reply by danieltott** [reply]
I think @clandau is correct, I should have used `parseInt` here.

But it only looks like magic because I'm not passing an anonymous function to `map` - I could rewrite it as this:

```js
// This:
["1", "2", "3"].map(Number); // [1, 2, 3]

// is the same as this:
["1", "2", "3"].map((str) => Number(str)); // [1, 2, 3]

// Just like this:
["1", "2", "3"].map((str) => parseInt(str)); // [1, 2, 3]

// Can be written like this:
["1", "2", "3"].map(parseInt); // [1, 2, 3]
```

Basically if your `map` input function takes one parameter and returns the result of another function that takes one parameter without doing anything to the parameter, you can just pass the target function directly.

**Reply by bogdaaamn** [reply]
> Basically if your map input function takes one parameter and returns the result of another function that takes one parameter without doing anything to the parameter, you can just pass the target function directly.

right right, sounds about right! thanks folks ❤️‍🔥

**Comment by clandau** [comment]
I'm shocked that it's day 4 and we haven't needed any crazy optimizations yet. Part 2 was more straightforward than 1 today imo

**Reply by clandau** [reply]
I apparently skipped last year, and went back and started it yesterday. it was harder earlier.

**Reply by doleraj** [reply]
I am still working through 2018, a year I gave up on Day 13, and Day 15 has had me stuck for days

**Reply by tkshill** [reply]
I've never made it past day 18 ever, and if I get to 10 this year without needing more than one day, I'll count that as a win.

But yeah, happy so far to not need any huge numbers or loop optimizations

**Comment by davidalpert** [comment]
fell asleep last night early before the problem dropped, then woke up and couldn't wait for @tkshill; had to get this one done before my day kicked off.

F# is really nice for working with sets and matrices as it comes with libraries that do all the heavy lifting and the paradigm is all about composing functions together to express the domain:

- [Day 04 - Camp Cleanup](https://github.com/davidalpert/advent-of-code-da/blob/bf87d42fcb450efb7e3f90eef358315c0ad8b3e4/2022/day04_camp_cleanup.fs#L11-L49) (as F#) - first solution

- [Day 04 - Camp Cleanup](https://github.com/davidalpert/advent-of-code-da/blob/2022/2022/day04_camp_cleanup.fs#L11-L62) (as F#) - refactored

I started with fewer pieces and wound up with more functions, more lines of code, but each function is smaller and simpler. maybe.

the first one seems easier to understand to me.

in the second one I can focus on [the solution](https://github.com/davidalpert/advent-of-code-da/blob/2022/2022/day04_camp_cleanup.fs#L50-L62) which is very small and very declarative, though in order to understand how it's implemented it you need to read many of the helpers...

**Reply by tkshill** [reply]
Ohhhhh, this is soo cool.

Comparing with an empty sequence is a little neater than my not empty check.

And acknowledging that both parts use that eitherOr constraint is genius!

I also forgot about `Seq.where` which as far as I can tell, is functionally the same as `Seq.filter`

**Reply by davidalpert** [reply]
I started with a not is empty check but it had a bug in it. I probably wrote it wrong. 🙈

**Reply by davidalpert** [reply]
I was actively thinking “what would Kirk do?”

**Reply by bogdaaamn** [reply]
classic WWKD

**Comment by danimeyer** [comment]
Ruby

I'm playing around with refactoring this one. Here's the commit with my original solution: https://github.com/danimeyer/advent-of-code/commit/5e56445248176abcc8d093348c21edbb74ae8bcd

And then here's the code in my repo. I want to try using some more ruby methods in my solution: https://github.com/danimeyer/advent-of-code/blob/main/2022/4.rb

**Reply by tkshill** [reply]
I'm always fascinated by how expressive ruby is. Also, watching your solution I finally see the overlap logic I was confusing myself about. Great solve!

**Comment by tkshill** [comment]
This timezone makes me feel like I'm always behind 😢 

Time to read your lovely solutions.

My Solution [here](https://github.com/tkshill/AdventOfCode2022/blob/main/AOC2022/Day04.fs). Came out pretty clean? Tried to figure out part 2 with _maths_ but said screw it, sets never fail me.

Took some time to debug, but shoutout to tests and scripting 😄 

Edit: Realizing now that both of these could've been done with sets or just comparisons, so I guess each of my parts shows one or the other.

**Reply by bogdaaamn** [reply]
I slowly start to be able to read F#

```fs
let isSubset ((l, r), right) =
    match right with
    | (l2, r2) when l2 >= l && r2 <= r -> 1
    | (l2, r2) when l2 <= l && r2 >= r -> 1
    | _ -> 0
```

This bit makes a lot of sense!

**Reply by Cerchie** [reply]
It's so beautiful!

**Reply by tkshill** [reply]
@BogDAAAMN If you can make heads or tails of this I appreciate you, I'm not writing it to be particularly readable, @davidalpert is being much more diligent.

And yeah, the language is reaaally good for pattern matching in a concise way

@Cerchie I like to think so. It's a style of programming where you try to like, make every word have meaning? There's definitely a point where you can be too terse, but it also just gives you less code to think about. But mainly just lots of function chaining. A to B to C. Once the types match you can just plug 'em in!

**Comment by turingtestable** [comment]
A very late [ruby](https://github.com/turingtestable/AdventOfCode2022/blob/master/challenge4/solution.rb) solution from me today. I spend a lot of time misunderstanding how the input data was coming in without actually looking at it. Also, today was tree decorating day so didn't get a chance to really finish thinking til now.

**Reply by bogdaaamn** [reply]
Very nice and clean solution, I like how you break down everything. Is it a ruby pattern or this is how you work? 

(Also literally just finished decorating the tree 🎄)

**Reply by turingtestable** [reply]
Generally just my brain, lol.

**Comment by bogdaaamn** [comment]
The quickest and obvious solution, still not optimising nothing 😭 

https://github.com/BogDAAAMN/advent-of-code-2022/blob/main/day-04.js

**Reply by Cerchie** [reply]
oooh but you use reduce 
![giphy (3)](https://user-images.githubusercontent.com/54046179/205518189-49de89f6-56aa-4694-a167-2ec9e15d411d.gif)

**Comment by Cerchie** [comment]
Yay, day1 in `index.js` and day2 in `index2.js`. I think I might not actually need the last `if else` block but I also just need a break 😭. : https://github.com/Cerchie/adventofcode2022/tree/main/day-04

**Reply by bogdaaamn** [reply]
Nice and clear, I like it!! 

What are the pluses in conditions for? 

```js
+firstArray[0] >= +secondArray[0] && +firstArray[1] <= +secondArray[1]
```

**Reply by Cerchie** [reply]
It coerces the string to a number in JS!

**Reply by bogdaaamn** [reply]
OHHH wow

**Reply by bogdaaamn** [reply]
Right, now I see you don't do any parsing in the beginning. Nicee and clean
